ducktape/utils/wait.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2015 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Conscious waiting
+#
+#
+#
+
+import time
+import traceback
+import os
+import errno
+import signal
+
+class TimeoutError (Exception):
+    pass
+
+time_saved_sum = 0
+
+def time_saved ():
+    """ Returns the amount of time saved using wait.until() instead of time.sleep() """
+    global time_saved_sum
+    return time_saved_sum
+
+
+def sleep (seconds):
+    """ Simple wrapper for time.sleep() but indicates that the programmer
+        made a conscious decisions not to use wait.until() """
+    return time.sleep(seconds)
+
+def until0 (timeout, expect_result, test_func, *args):
+    """ Wait up to 'timeout' seconds for 'test_func(*args)' to return 'expect_result'.
+        Raises TimeoutError() when 'timeout' is reached, else True if function
+        returned expected results in time.
+        'test_func(*args)' is called repeatedly for the duration of the 'timeout' period,
+        with a small pause between each invocation to avoid busy looping.
+    """
+    start_time = time.time()
+    abs_tmout = start_time + timeout
+
+    interval = timeout / 50.0
+
+    while time.time() < abs_tmout:
+        r = False
+        try:
+            r = test_func(*args)
+        except:
+            traceback.print_exc()
```

Comment:
This goes to stdout -- shouldn't it be going to some logger?

I think we've struggled a bit with this in other utilities (e.g. we have to set a logger on `RemoteAccount` and what we decide to log and at what level has been pretty ad hoc and has only kind of worked so far).

Can we choose one of these options?:
1. Treat exceptions as an accepted failure mode and don't log anything. In other words, callers should expect that *any* exception thrown by the test_func will simply be treated as not meeting the condition. (Note that this gets a bit weird for `until_not`, which could exit due to an unrelated exception rather than bec).
2. Don't handle the exception at all. The caller should be implementing something that actually returns True or False, and if it throws an exception we consider that an error and say we can't be sure of what it means (by letting it propagate up the stack).